python/ray/data/impl/batcher.py

```diff
@@ -63,7 +63,7 @@ def next_batch(self) -> Block:
                 leftover.append(block)
             elif accessor.num_rows() <= needed:
                 # We need this entire block to fill out a batch.
-                output.add_block(block)
+                output.add_block(accessor.slice(0, accessor.num_rows(), copy=False))
```

Comment:
Can you explain why is this different than just passing in the block? It looks identical to me.

Reply:
In my opinion, **because the type of `block` obtained through `test_raydp_to_torch_iter()` is `bytes`**, then in the source code, the  [accessor](https://github.com/ray-project/ray/blob/cc0db8b92addc323eb1f58060d58d343b0bd9225/python/ray/data/impl/batcher.py#L59) created here is `ArrowBlockAccessor`, and then pass https://github.com/ray-project/ray/blob/cc0db8b92addc323eb1f58060d58d343b0bd9225/python/ray/data/impl/batcher.py#L66 reach https://github.com/ray-project/ray/blob/cc0db8b92addc323eb1f58060d58d343b0bd9225/python/ray/data/impl/delegating_block_builder.py#L36-L44 The `accessor` created here is still `ArrowBlockAccessor`, **similarly the `builder` created by Line43 is `ArrowBlockBuilder`**.  So when it goes through Line44 to https://github.com/ray-project/ray/blob/cc0db8b92addc323eb1f58060d58d343b0bd9225/python/ray/data/impl/table_block.py#L48 **the type of `self._block_type` is `pyarrow` , but the type of `block` is still `bytes`, which will hit an assertion error**.
So I think using `accessor.slice()` instead of `block` directly could solve this problem.

Reply:
Thank you Keping for chasing down the details!

Reply:
Thanks for figuring out the root cause! Could you add a comment summarizing the issue, so it's obvious why this seemingly no-op slice is necessary?

Also, this is making me think that we need a much better story around generic bytes handling.

Reply:
Thans a lot, I've done.

Reply:
Ah @KepingYan I meant add a comment to the code, where this slice takes place, so it's obvious for devs why this no-op slice is necessary without having to search through git history.

Reply:
I got it!